pulser/simulation/simulation.py

```diff
@@ -325,55 +463,125 @@ def get_hamiltonian(self, time: float) -> qutip.Qobj:
-                evolution. Will be transformed into a ``qutip.Qobj`` instance.
-            progress_bar (bool): If True, the progress bar of QuTiP's
-                ``qutip.sesolve()`` will be shown.
+        Will return NoisyResults if it detects any noise in the SimConfig.
+        Otherwise will return CleanResults.
 
-        Other Parameters:
-            options: Additional simulation settings. These correspond to the
-                keyword arguments of ``qutip.solver.Options`` for
-                the ``qutip.sesolve()`` method.
-
-        Returns:
-            SimulationResults: Object containing the time evolution results.
+        Keyword Args:
+            progress_bar (bool): If True, the progress bar of QuTiP's solver
+                will be shown.
+            options (qutip.solver.Options): If specified, will override
+                SimConfig solver_options.
         """
-        if initial_state is not None:
-            if isinstance(initial_state, qutip.Qobj):
-                if initial_state.shape != (self.dim**self._size, 1):
-                    raise ValueError("Incompatible shape of initial_state")
-                self._initial_state = initial_state
+        solv_ops = qutip.Options(max_step=5, **options) if options \
+            else self.config.solver_options
+
+        def _assign_meas_basis() -> str:
+            if hasattr(self._seq, '_measurement'):
+                return cast(str, self._seq._measurement)
             else:
-                if initial_state.shape != (self.dim**self._size,):
-                    raise ValueError("Incompatible shape of initial_state")
-                self._initial_state = qutip.Qobj(initial_state)
-        else:
-            # by default, initial state is "ground" state of g-r basis.
-            all_ground = [self.basis['g'] for _ in range(self._size)]
-            self._initial_state = qutip.tensor(all_ground)
-
-        result = qutip.sesolve(self._hamiltonian,
-                               self._initial_state,
-                               self.eval_times,
-                               progress_bar=progress_bar,
-                               options=qutip.Options(max_step=5,
-                                                     **options)
-                               )
-        meas_basis: Optional[str]
-        if hasattr(self._seq, '_measurement'):
-            meas_basis = cast(str, self._seq._measurement)
+                if self.basis_name in {'digital', 'all'}:
+                    return 'digital'
+                else:
+                    return 'ground-rydberg'
+
+        def _run_solver(as_subroutine: bool = False,
+                        measurement_basis: str = '') -> CleanResults:
+            """Returns CleanResults: Object containing evolution results."""
+            if not as_subroutine:
+                # CLEAN SIMULATION:
+                measurement_basis = _assign_meas_basis()
+            if self._collapse_ops:
```

Comment:
This should check if `dephasing` is in `self.config.noise` instead. I don't think you reset `self._collapse_ops` when you reset config, but you shouldn't have to either if you change this

Reply:
Agreed, but in the future we might have other noises that rely on the density matrix formalism. But when time comes this'll be added. I'll makes the changes